tooling/nargo/src/ops/execute.rs

```diff
@@ -184,25 +194,70 @@ impl<'a, F: AcirField, B: BlackBoxFunctionSolver<F>, E: ForeignCallExecutor<F>>
         // included in a failure case.
         self.call_stack.clear();
 
-        Ok(acvm.finalize())
+        if self.profiling_active {
+            Ok(acvm.finalize_with_profiling())
+        } else {
+            Ok((acvm.finalize(), ProfilingSamples::default()))
```

Comment:
I think this double-checking of which version of finalize to call, only then to return an empty list, which then again is double checked to be empty, is a bit too defensive. I reckon an `Option` which is `None` if `profiling_active` is `false` should be enough, we could trust the ACVM to do the right thing, and then all the subsequent processing becomes more streamlined.

Reply:
`None` for the second tuple field of the profiling samples? This feels partially redundant with knowing that we would expect no samples when profiling is not active.

Reply:
Sorry, my comment was confusing because the code involved is a spread out. 

What I meant was that this pattern:
```
struct VM { profiling: bool, samples: Samples }
impl VM {
  fn finalize() -> Witness { 
    assert!(!self.profiling);
    ...
  }
  fn finalize_with_profiling(self) -> (Witness, Samples) {
    assert!(self.profiling);
    ...
  }
}

fn foo(profiling: bool) -> (Witness, Samples) {
  let vm = VM::new(..., profiling = self.profiling);
  ...
  let (result, samples) = if self.profiling { 
    vm.finalize_with_profiling()
  } else { 
    (vm.finalize(), Samples::default())
  };
}

fn bar() -> Witness {
  let (result, samples) = foo(false);
  assert(samples.is_empty());
  result
}

fn baz() -> (Witness, Samples) {
  foo(true)
}
```
is like realising the four eyes principle in code, with the caller and the callee constantly second guessing each other. Like there are samples in VM, but they aren't returned, but then a default is added, which is asserted to be empty, and then discarded...

I just meant that if we trust that the VM itself will return `Some(Samples)` if `profiling_active` is `true` and `None` if it's false, then the rest of them can be without further assertions. 

I would be in support of an API which looks like `let (witness, sample) = VM::new(...).with_profiling().finalize()` or `let witness = VM::new().finalize()` (ie. a wrapper) but that doesn't seem possible because the Solver also just receives a flag from the outside and has to handle both cases. I'm just saying the double checking seems unnecessary. Maybe a `.expect("profiling is on")` by a caller that passed `true` and really wants those samples
 enough?

Reply:
I understand the motivation that you don't want to accidentally profile when the code isn't even consuming it, just seemed like it's negated here a bit with the re-introduction of a default value.

Reply:
Ok I see what you mean. And it is a fair point that I should trust the ACVM to act appropriately. I switched to just fetching the profiling samples before calling `finalize`. I then actually check whether the profiling samples returned from the ACVM are empty when I am not profiling rather than that default value.